File: dev/set-store-tests.py
#!/usr/bin/env python3
"""Set destinations, aliasing, atomic errors and parity with local Redis."""
import importlib.util
import json
from pathlib import Path
import subprocess
import sys
import tempfile

ROOT = Path(__file__).resolve().parents[1]
spec = importlib.util.spec_from_file_location('set_store_algebra', ROOT / 'dev/set-algebra-tests.py')
algebra = importlib.util.module_from_spec(spec)
spec.loader.exec_module(algebra)
arrays, support = algebra.arrays, algebra.support
run, require, literal = support.run, support.require, support.literal
TAG = '(tag b"store")'
KEYS = tuple('{store}:teams:' + key for key in ('destination', 'left', 'right'))
KINDS = ('union', 'inter', 'diff')
SUFFIXES = ('', 'Left', 'Right', 'Same', 'AllSame', 'Reverse', 'Within', 'Earlier', 'Typed', 'Snapshot', 'Raw')
ENTRIES = tuple(kind + suffix for kind in KINDS for suffix in SUFFIXES)
EXAMPLES = {'main': b'["alice","bob","carol"]\n', 'retained': b'1\n', 'exclusive': b'["alice"]\n'}
INITIAL = ({b'old'}, {b'a', b'c'}, {b'b', b'c'})


def operands(suffix):
    destination = 1 if suffix in ('Left', 'AllSame') else 2 if suffix == 'Right' else 0
    left, right = (2, 1) if suffix == 'Reverse' else (1, 1) if suffix in ('Same', 'AllSame') else (1, 2)
    return destination, left, right


def case(kind, suffix, before):
    destination, left, right = operands(suffix)
    if any(value is not None and not isinstance(value, set) for value in (before[left], before[right])):
        return kind + suffix, before, support.WRONG, before, 'status', False
    members = algebra.operation(kind, before[left], before[right])
    after = list(before)
    after[destination] = set(members) if members else None
    reply, reply_kind = str(len(members)).encode(), 'bulk'
    if suffix in ('Within', 'Earlier'):
        after[destination] = None
    elif suffix == 'Typed':
        reply, reply_kind = b'integer', 'status'
    elif suffix == 'Snapshot':
        reply, reply_kind = members, 'array'
    return kind + suffix, before, reply, tuple(after), reply_kind, True


def cases():
    inputs = [(None, None), ({b'a', b'b'}, None), (None, {b'a', b'b'}), INITIAL[1:],
              ({b'a'}, {b'b'}), ({b'a\0', b'a', b''}, {b'aa', b'a', b'\0'}),
              ({b'10', b'2'}, {b'01', b'2'}), ({support.TEXT, b''}, {'café'.encode(), b''}),
              ({bytes([i]) for i in range(256)}, {b'', b'\xff'}),
              ({f'm{i:03}'.encode() for i in range(129)}, {f'm{i:03}'.encode() for i in range(65)})]
    rows = []
    for kind in KINDS:
        rows += [case(kind, '', ({b'f': b'v'}, *pair)) for pair in inputs]
        rows += [case(kind, '', (old, *INITIAL[1:])) for old in (None, {b'old'}, b'old', [b'old'])]
        rows += [case(kind, suffix, (INITIAL[0], *pair)) for suffix in ('Left', 'Right') for pair in inputs[:5]]
        for bad in (b'wrong', {b'f': b'v'}, [b'm']):
            for good in (None, {b'a'}):
                for pair in ((bad, good), (good, bad)):
                    rows += [case(kind, suffix, (INITIAL[0], *pair)) for suffix in ('', 'Left', 'Right')]
        rows += [case(kind, suffix, INITIAL) for suffix in ('Same', 'AllSame', 'Reverse', 'Within', 'Earlier', 'Typed', 'Snapshot')]
    require(len(rows) == 201, 'SET-STORE case inventory')
    return rows


def source(entry):
    kind = next(k for k in KINDS if entry.startswith(k))
    suffix = entry[len(kind):]
    destination, left, right = operands(suffix)
    keys = tuple(f'(teams b"{key}")' for key in ('destination', 'left', 'right'))
    command = f's{kind}store Reply {TAG} {keys[destination]} {keys[left]} {keys[right]}'
    declarations = ['module SetStoreCases', 'schema teams : String -> Key Set tag b"store"',
        'def expose : Reply -> Reply := fun (r : Reply) => case r as x in Reply return Reply with',
        '| nil => nil | int n => int n | bulk b => bulk b',
        '| status b => status b | err b => status b | array rs => array rs',
        'def typed : Reply -> Reply := fun (r : Reply) => case r as x in Reply return Reply with',
        '| int n => status b"integer" | nil => nil | bulk b => status b"wrong"',
        '| status b => status b"wrong" | err b => status b | array rs => status b"wrong"',
        f'def removeScript : Script Reply {TAG} := do {{ _ <- del Reply Set {TAG} {keys[destination]}; pure Reply {TAG} nil }}']
    tail = f'_ <- del Reply Set {TAG} {keys[destination]}; ' if suffix == 'Within' else ''
    if suffix == 'Snapshot':
        tail += f'members <- smembers Reply {TAG} {keys[destination]}; '
    answer = 'r' if suffix in ('Raw', 'Within') else 'typed r' if suffix == 'Typed' else 'members' if suffix == 'Snapshot' else 'expose r'
    declarations += [f'def {entry}Script : Script Reply {TAG} := do {{ r <- {command}; {tail}pure Reply {TAG} ({answer}) }}']
    later = f'_ <- inv Reply {TAG} removeScript; ' if suffix in ('Earlier', 'Raw') else ''
    declarations += [f'def {entry} : Client Reply := do {{ r <- inv Reply {TAG} {entry}Script; {later}done Reply r }}']
    return '\n'.join(declarations) + '\n'


def emit(work, entry):
    path, output = work / 'SetStoreCases.tet', work / entry
    path.write_text(source(entry))
    run(['./tether', 'emit', str(path), '--entry', entry, '-o', str(output)])
    require(b'LUA-SAME ' in run(['node', 'dev/lua-same.mjs', str(output)]), 'SET-STORE canonical bodies')
    suffix = entry[len(next(k for k in KINDS if entry.startswith(k))):]
    destination, left, right = operands(suffix)
    for script in json.loads((output / 'scripts.json').read_text()):
        body = (output / (script['stem'] + '.lua')).read_bytes()
        require(body.startswith(b'#!lua\n'), 'SET-STORE write classification')
        indices = (destination,) if script['entry'] == 'removeScript' else (destination, left, right)
        require(script['keys'] == sorted({KEYS[i] for i in indices}), 'SET-STORE declared keys')
    return output


def twin(work, output, row):
    _entry, before, expected, after, kind, _success = row
    algebra.twin(work, output, before, expected, after, kind, keys=KEYS, reason='SET-STORE LuaJIT reply')


def refusals(work):
    invalid = []
    mismatch = b'CHECK mismatch: the constructor keyBytes of Key gives the index '
    for kind in KINDS:
        for side in range(3):
            for schema, ctor in [('(Str Binary)', 'Str'), ('Hash', 'Hash'), ('List', 'List'), ('ZSet', 'ZSet'), ('Stream', 'Stream')]:
                types, tags = ['Set'] * 3, ['store'] * 3
                types[side] = schema
                invalid.append((kind, types, tags, mismatch + f'(In SMu RedisType [] (ACtor {ctor})'.encode()))
            tags = ['store'] * 3
            tags[side] = 'elsewhere'
            invalid.append((kind, ['Set'] * 3, tags, mismatch + b'(In SMu Tag [] (ACtor tag)'))
    for index, (kind, types, tags, diagnostic) in enumerate(invalid):
        path, output = work / 'Refused.tet', work / f'refused-{index}'
        path.write_text('module Refused\n' + '\n'.join(
            f'schema {name} : String -> Key {schema} tag b"{tag}"' for name, schema, tag in zip(('destination', 'left', 'right'), types, tags)) +
            f'\ndef s : Script Reply {TAG} := do {{ r <- s{kind}store Reply {TAG} (destination b"x") (left b"x") (right b"x"); pure Reply {TAG} r }}\n'
            f'def main : Client Reply := do {{ r <- inv Reply {TAG} s; done Reply r }}\n')
        result = subprocess.run(['./tether', 'emit', str(path), '-o', str(output)], cwd=ROOT, capture_output=True, timeout=120)
        require(result.returncode == 2 and diagnostic in result.stderr, f'SET-STORE refusal {index}: {result.stderr!r}')
        require(not output.exists(), f'SET-STORE refusal published output {index}')
    require(len(invalid) == 54, 'SET-STORE refusal inventory')
    print(f'PASS SET-STORE-REFUSALS cases={len(invalid)} atomic_output={len(invalid)}', flush=True)


def live(work, outputs):
    extra = []
    for kind in KINDS:
        for bad in ('zset', 'stream'):
            extra.append(case(kind, '', (bad, *INITIAL[1:])))
            for good in (None, {b'a'}):
                for pair in ((bad, good), (good, bad)):
                    extra += [case(kind, suffix, (INITIAL[0], *pair)) for suffix in ('', 'Left', 'Right')]
    members_hex = "local o={}; for _,v in ipairs(redis.call('SMEMBERS',KEYS[1])) do o[#o+1]=(v:gsub('.',function(c) return string.format('%02x',string.byte(c)) end)) end; table.sort(o); return table.concat(o,',')"
    with support.local.hosts(work) as (port, env):
        redis = ['redis-cli', '-h', '127.0.0.1', '-p', str(port), '--raw']
        run(redis + ['ACL', 'SETUSER', 'default', '-eval_ro', '-evalsha_ro'])
        hosts = errors = 0
        for row in cases() + extra:
            entry, initial, expected, after, reply_kind, success = row
            suffix = entry[len(next(k for k in KINDS if entry.startswith(k))):]
            destination, _left, _right = operands(suffix)
            for host in ('node', 'bash'):
                run(redis + ['FLUSHDB'])
                run(redis + ['SET', 'other', 'kept'])
                for key, value in zip(KEYS, initial):
                    algebra.seed(redis, key, value)
                    run(redis + ['EXPIRE', key, '600'])
                snapshots = [run(redis + ['DUMP', key]) for key in KEYS]
                output = outputs[entry]
                args = (['node', 'runtime/redis-host.mjs', str(output / 'prog.wasm'), str(port)]
                        if host == 'node' else ['/bin/bash', str(output / 'prog.sh')])
                require(run(args, env=env) == arrays.stdout(expected, reply_kind), f'SET-STORE {host} {entry} reply')
                for index, (key, previous, final, snapshot) in enumerate(zip(KEYS, initial, after, snapshots)):
                    if not success or index != destination:
                        require(run(redis + ['DUMP', key]) == snapshot, 'SET-STORE preserved complete value')
                        expiry = run(redis + ['TTL', key])
                        require(int(expiry) > 0 if previous is not None else expiry == b'-2\n',
                                'SET-STORE preserved expiry')
                    elif final is None:
                        require(run(redis + ['EXISTS', key]) == b'0\n', 'SET-STORE empty result deletes destination')
                    else:
                        require(run(redis + ['TYPE', key]) == b'set\n', 'SET-STORE destination type')
                        require(run(redis + ['SCARD', key]) == str(len(final)).encode() + b'\n', 'SET-STORE destination cardinality')
                        require(run(redis + ['EVAL', members_hex, '1', key]) == b','.join(sorted(v.hex().encode() for v in final)) + b'\n',
                                'SET-STORE destination members')
                        require(run(redis + ['TTL', key]) == b'-1\n', 'SET-STORE replacement clears expiry')
                require(run(redis + ['GET', 'other']) == b'kept\n', 'SET-STORE unrelated key')
                hosts += 1
        for kind in KINDS:
            for host in ('node', 'bash'):
                run(redis + ['FLUSHDB'])
                run(redis + ['SET', KEYS[0], 'kept'])
                run(redis + ['SET', KEYS[2], 'wrong'])
                output = outputs[kind + 'Raw']
                args = (['node', 'runtime/redis-host.mjs', str(output / 'prog.wasm'), str(port)]
                        if host == 'node' else ['/bin/bash', str(output / 'prog.sh')])
                require(run(args, env=env, code=4) == b'', 'SET-STORE error stops host')
                require(run(redis + ['GET', KEYS[0]]) == b'kept\n', 'SET-STORE error stops later invocation')
                require(run(redis + ['GET', KEYS[2]]) == b'wrong\n', 'SET-STORE error preserves source')
                hosts, errors = hosts + 1, errors + 1
        require((len(extra), hosts, errors) == (78, 564, 6), 'SET-STORE host inventory')
        print(f'PASS SET-STORE-E2E cases=279 hosts={hosts} errors={errors}', flush=True)
    count = 0
    for entry, expected in EXAMPLES.items():
        for host in ('node', 'bash', 'luajit'):
            require(run(['./tether', 'exec', 'examples/TeamCache.tet', '--entry', entry, '--host', host]) == expected,
                    f'SET-STORE example {entry} {host}')
            count += 1
    print(f'PASS SET-STORE-EXAMPLE exec={count}', flush=True)


def main():
    probe = len(sys.argv) == 3 and sys.argv[1] == '--probe'
    require(probe or sys.argv[1:] in ([], ['--static'], ['--offline']),
            'Usage: set-store-tests.py [--static|--offline|--probe ENTRY]')
    with tempfile.TemporaryDirectory(prefix='tether-set-store-') as temporary:
        work = Path(temporary)
        if probe:
            entry = sys.argv[2]
            rows = [row for row in cases() if row[0] == entry]
            require(bool(rows), 'SET-STORE probe has no cases')
            output = emit(work, entry)
            for row in rows:
                twin(work, output, row)
            print(f'PASS SET-STORE-PROBE entry={entry} cases={len(rows)}', flush=True)
            return
        outputs = {entry: emit(work, entry) for entry in ENTRIES}
        print(f'PASS SET-STORE-ARTIFACTS pairs={len(outputs)}', flush=True)
        refusals(work)
        for entry, expected in EXAMPLES.items():
            require(run(['./tether', 'run', 'examples/TeamCache.tet', '--entry', entry]) == expected, 'SET-STORE interpreter example')
        print(f'PASS SET-STORE-STORE-EXAMPLE cases={len(EXAMPLES)}', flush=True)
        if '--static' not in sys.argv:
            for row in cases():
                twin(work, outputs[row[0]], row)
            print(f'PASS SET-STORE-ORACLES luajit={len(cases())}', flush=True)
        if not sys.argv[1:]:
            live(work, outputs)
    mode = f" mode={sys.argv[1].lstrip('-')}" if sys.argv[1:] else ''
    print(f'PASS SET-STORE-TESTS{mode}', flush=True)


if __name__ == '__main__':
    try:
        main()
    except (AssertionError, OSError, ValueError, subprocess.SubprocessError) as error:
        print(f'FAIL SET-STORE-TESTS {error}', file=sys.stderr)
        sys.exit(1)

File: dev/set_store_tests.ml
module S = Tether_store.Store
module I = Tether_store.Interp
module E = Kanon_kernel.Eterm
let ( let* ) = Result.bind
let require condition reason = if condition then Ok () else Error ("SET-STORE-UNIT " ^ reason)
let fold f xs = List.fold_left (fun acc x -> let* n = acc in let* () = f x in Ok (n + 1)) (Ok 0) xs
let same a b = S.Keys.bindings a = S.Keys.bindings b
let rec term = function
  | I.Data (tid, tag, xs) -> E.KTag (tid, tag, List.map term xs)
  | I.Literal l -> E.KLit l
  | I.Fields _ | I.Closure _ | I.Erased -> E.KErased
let bytes s = term (I.bytes s)
let key s = E.KTag (E.Tid "mu<Key>", 0, [bytes s])
let execute ?(expose=false) ?(destination="destination") tag args before =
  let script = E.KTag (E.Tid "mu<Script>", tag, key destination :: args @ [E.KClos (E.Fid "pure", 1, [])]) in
  let client = E.KTag (E.Tid "mu<Client>", 1, [script; E.KClos (E.Fid "done", 1, [])]) in
  let fn name ps body = E.KFun (E.Fid name, ps, E.RI31, body) in
  let answer = if expose then E.KCase (E.Tid "mu<Reply>", E.KVar 0,
    [{E.tag=4; arity=1; body=E.KTag (E.Tid "mu<Reply>", 3, [E.KVar 0])}]) else E.KVar 0 in
  let rows = ["main", Kanon_kernel.Erase.Code [fn "main" [] client;
    fn "pure" [E.RI31] (E.KTag (E.Tid "mu<Script>", 0, [answer]));
    fn "done" [E.RI31] (E.KTag (E.Tid "mu<Client>", 0, [E.KVar 0]))]] in
  I.run ~budget:(Kanon_kernel.Budget.of_poll (fun () -> false)) rows ~entry:"main" before
let destinations = ["destination"; "left"; "right"]
let wrong = [S.Str "old"; S.Hash ["f", "v"]; S.List ["m"]; S.ZSet ["m", "1"]; S.Stream ["entry", "v"]]
let initial = S.empty |> S.put "other" (S.Str "kept") |> S.put "destination" (S.Hash ["stale", "value"])
let run () =
  let octets = List.of_seq (String.to_seq I.octets) |> List.map (String.make 1) in
  let numbered first count = List.init count (fun i -> Printf.sprintf "m%03d" (first + i)) in
  let cases = [
    None, None, [], [], [], "missing";
    Some ["b"; "a"], None, ["a"; "b"], [], ["a"; "b"], "left only";
    None, Some ["b"; "a"], ["a"; "b"], [], [], "right only";
    Some ["c"; "a"], Some ["c"; "b"], ["a"; "b"; "c"], ["c"], ["a"], "overlap";
    Some ["b"], Some ["a"], ["a"; "b"], [], ["b"], "disjoint";
    Some ["b"; "a"; "a"], Some ["b"; "b"], ["a"; "b"], ["b"], ["a"], "duplicates";
    Some ["a\000"; "a"; ""], Some ["aa"; "a"; "\000"],
      [""; "\000"; "a"; "a\000"; "aa"], ["a"], [""; "a\000"], "prefix bytes";
    Some ["10"; "2"], Some ["01"; "2"], ["01"; "10"; "2"], ["2"], ["10"], "decimal bytes";
    Some (List.rev octets), Some [""; "\255"], "" :: octets, ["\255"],
      List.filter (fun s -> s <> "\255") octets, "all bytes";
    Some (List.rev (numbered 0 129)), Some (numbered 0 65), numbered 0 129,
      numbered 0 65, numbered 65 64, "complete set"] in
  let* total = List.fold_left (fun acc tag -> let* total = acc in
    let check destination args before expected name =
      let* answer, after = execute ~destination tag args before in
      let wanted = if expected = [] then S.Keys.remove destination before else S.put destination (S.Set expected) before in
      require (answer = I.Int (string_of_int (List.length expected)) && same after wanted) name in
    let fixtures = List.concat_map (fun row -> List.map (fun d -> d, row) destinations) cases in
    let* values = fold (fun (destination, (left, right, union, inter, diff, name)) ->
      let put k value st = Option.fold ~none:st ~some:(fun xs -> S.put k (S.Set xs) st) value in
      let before = initial |> put "left" left |> put "right" right in
      check destination [key "left"; key "right"] before
        (if tag = 35 then union else if tag = 36 then inter else diff) name) fixtures in
    let* replaced = fold (fun old ->
      let before = Option.fold ~none:(S.Keys.remove "destination" initial) ~some:(fun d -> S.put "destination" d initial) old
        |> S.put "left" (S.Set ["a"; "b"]) |> S.put "right" (S.Set ["b"; "c"]) in
      check "destination" [key "left"; key "right"] before
        (if tag = 35 then ["a"; "b"; "c"] else if tag = 36 then ["b"] else ["a"]) "overwrite destination")
      (None :: List.map Option.some (S.Set ["old"] :: wrong)) in
    let errors = List.concat_map (fun bad -> List.concat_map (fun side -> List.concat_map (fun present ->
      List.map (fun d -> bad, side, present, d) destinations) [false; true]) ["left"; "right"]) wrong in
    let* errors = fold (fun (bad, side, present, destination) ->
      let other = if side = "left" then "right" else "left" in
      let before = S.put side bad initial in
      let before = if present then S.put other (S.Set ["a"]) before else before in
      let args = [key "left"; key "right"] in
      let* () = require (execute ~destination tag args before = Error (S.message S.Wrong_type)) "error stops client" in
      let* answer, after = execute ~expose:true ~destination tag args before in
      require (answer = I.Status (S.message S.Wrong_type) && same before after) "error preserves destination") errors in
    let* identical = fold (fun destination ->
      let before = S.put "left" (S.Set ["b"; "a"]) initial in
      check destination [key "left"; key "left"] before (if tag = 37 then [] else ["a"; "b"]) "same source") destinations in
    let signed = E.KTag (E.Tid "mu<Signed64>", 0, [bytes "0"]) in
    let* shapes = fold (fun args -> require (execute tag args initial = Error "STORE-SCRIPT-COMMAND") "operand shape")
      [[]; [key "left"]; [bytes "left"; key "right"]; [key "left"; bytes "right"];
       [signed; key "right"]; [key "left"; signed]; [key "left"; key "right"; key "other"]] in
    Ok (total + values + replaced + errors + identical + shapes)) (Ok 0) [35; 36; 37] in
  let* () = require (execute 39 [key "left"; key "right"] initial = Error "STORE-SCRIPT-COMMAND") "unknown command" in
  let total = total + 1 in
  let* () = require (total = 322) (Printf.sprintf "expected 322 cases, counted %d" total) in Ok total
let () = run () |> Result.fold
  ~ok:(fun count -> Printf.printf "PASS SET-STORE-UNIT cases=%d\n" count)
  ~error:(fun e -> prerr_endline ("FAIL " ^ e); exit 1)
